### PR TITLE
Added new Services view (FATE#321043)

### PR DESCRIPTION
**Pull request by lslezak**
## :memo: Notes

- I originally started the new "Services" view as a copy of the already existing "Repositories" view, but later I did some refactoring and sharing some common functionality - maybe reviewing commits one-by-one is easier than checking the whole change at once.
- The NCurses part is implemented [here](https://github.com/libyui/libyui-ncurses-pkg/pull/19)
- The issue with underscores in the service names is reported separately at https://bugzilla.suse.com/show_bug.cgi?id=1099851 (it's a yast2-registration issue)

## :sparkles: Features

- Added new "Services" view to the filter list, see [FATE#321043](https://fate.suse.com/321043) for more details
- The view is only available if at least one service is present (so usually not visible in openSUSE or unregistered SLE)
- The secondary filter also works, e.g. you can search packages in the selected service, just like in the original "Repositories" view
- Display the busy cursor while the package filtering is in progress. The package filtering might take some time if there are many packages. It is now also used by the other views like the repository view.
- The service list is sorted case insensitive, that's why the `SUSE_Linux_Enterprise_15_x86_64` service is below the `Server_Applications_Module_x86_64`. That's a difference from the original "Repositories" view which uses the case sensitive sort.
- We should probably also change this in the original "Repositories" view but probably only in SP1 as this could potentially break something in openQA.

## :world_map: Screenshot

qt-pkg-services

### CODE REVIEW

File: src/YQPkgSecondaryFilterView.h [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+     * Constructor
+     **/
+    YQPkgSecondaryFilterView( QWidget * parent );
+
+    /**
+     * Destructor
+     **/
+    virtual ~YQPkgSecondaryFilterView();
+
+    void init(QWidget * primary_widget);
```
Comment (mvidner) [inline src/YQPkgSecondaryFilterView.h]: is the argument really a generic widget? the implementation seems to expect some specific signals from it. please add a bit of documentation.

File: src/YQPkgServiceList.h [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+namespace zypp {
+    class RepoManager;
+}
+
+// just a service name
+typedef std::string	ZyppService;
+
+
+/**
+ * @short Display a list of zypp::Selection objects.
```
Comment (mvidner) [inline src/YQPkgServiceList.h]: really zypp::Selection?

File: src/YQPkgRepoFilterView.h [hunk 3]
```diff
@@ -73,89 +67,14 @@ class YQPkgRepoFilterView : public QWidget
...
-     **/
-    bool secondaryFilterMatch( ZyppSel 	selectable,
-			       ZyppPkg 		pkg );
-
+    virtual void primaryFilter();
+    virtual void primaryFilterIfVisible();
 
     // Data members
-
     YQPkgRepoList *		_repoList;
```
Comment (mvidner) [inline src/YQPkgRepoFilterView.h]: Please document who owns the pointer. AFAICT it gets passed to the  `parent` QObject that becomes its owner (so this class cannot own it via a smart pointer). So `// owned by my parent` ?

File: src/YQPkgServiceList.h (hunk 2 above)
Comment (lslezak) [reply src/YQPkgServiceList.h]: Oh, copy&paste error, actually the problem is already in the original place...

File: src/YQPkgSecondaryFilterView.cc [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+    QHBoxLayout *layout = new QHBoxLayout(this);
+    layout->setContentsMargins(0,0,0,0);
+
+    QSplitter * splitter = new QSplitter( Qt::Vertical, this );
+    YUI_CHECK_NEW( splitter );
+
+    layout->addWidget( splitter );
+    splitter->addWidget(primary_widget);
+
+    // YUI_CHECK_NEW( _serviceList );
```
Comment (shundhammer) [inline src/YQPkgSecondaryFilterView.cc]: NP: leftover cruft?

File: src/YQPkgSecondaryFilterView.cc [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+
+
+YQPkgSecondaryFilterView::YQPkgSecondaryFilterView( QWidget * parent )
+    : QWidget( parent )
+{
+}
+
+void YQPkgSecondaryFilterView::init(QWidget * primary_widget)
+{
+    QHBoxLayout *layout = new QHBoxLayout(this);
```
Comment (shundhammer) [inline src/YQPkgSecondaryFilterView.cc]: No `YUI_CHECK_NEW(layout)` ?

File: src/YQPkgSecondaryFilterView.cc [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+
+    //
+    // All Packages
+    //
+    _allPackages = new QWidget( this );
+    YUI_CHECK_NEW( _allPackages );
+    _secondaryFilters->addPage( _( "All Packages" ), _allPackages );
+
+
+    // Unmaintaned packages: Packages that are not provided in any of
```
Comment (shundhammer) [inline src/YQPkgSecondaryFilterView.cc]: Typo: Unmainta**i**ned

File: src/YQPkgSecondaryFilterView.cc [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+             this,              &YQPkgSecondaryFilterView::filter );
+
+    //
+    // Status change view
+    //
+    _statusFilterView = new YQPkgStatusFilterView( parent );
+    YUI_CHECK_NEW( _statusFilterView );
+    _secondaryFilters->addPage( _( "Installation Summary" ), _statusFilterView );
+
+    connect( _statusFilterView,	SIGNAL( filterStart() ),
```
Comment (shundhammer) [inline src/YQPkgSecondaryFilterView.cc]: NP: Inconsistent indentation `SIGNAL()` vs. `SLOT()`

File: src/YQPkgSecondaryFilterView.h [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+class YQPkgSecondaryFilterView : public QWidget
+{
+    Q_OBJECT
+
+public:
+
+    /**
+     * Constructor
+     * Because of the circular dependencies you have to call the init()
+     * methot later with the new subclassed object.
```
Comment (shundhammer) [inline src/YQPkgSecondaryFilterView.h]: Typo: method

File: src/YQPkgServiceFilterView.h [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+#ifndef YQPkgServiceFilterView_h
+#define YQPkgServiceFilterView_h
+
+#include "YQPkgSecondaryFilterView.h"
+
+class QWidget;
+class YQPkgServiceList;
+
+class YQPkgServiceFilterView : public YQPkgSecondaryFilterView
```
Comment (shundhammer) [inline src/YQPkgServiceFilterView.h]: Class documentation missing

File: src/YQPkgServiceList.cc [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+
+YQPkgServiceList::YQPkgServiceList( QWidget * parent )
+    : QY2ListView( parent )
+{
+    yuiDebug() << "Creating service list" << endl;
+
+    QStringList headers;
+
+    // Column headers for repository list
+    headers <<  _("Name");
```
Comment (shundhammer) [inline src/YQPkgServiceList.cc]: Hint: If the margins in the header don't look good, it's typically a good approach to add a blank to the left and the right of the string - outside the translations, that is. I.e.

    `headers << ( QString(" ") + _("Name") + "" );`

This reserves a little space to the left and right. You can also achieve this by fiddling with margins in the QHeader, but that's a PITA; for the future, I recommend just adding a blank on each side.

File: src/YQPkgServiceList.cc [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+
+    setHeaderLabels( headers );
+    header()->setSectionResizeMode( _nameCol, QHeaderView::Stretch );
+
+    //setAllColumnsShowFocus( true );
+    setSelectionMode( QAbstractItemView::ExtendedSelection );	// allow multi-selection with Ctrl-mouse
+
+    connect( this, 	SIGNAL( itemSelectionChanged() ),
+	     this, 	SLOT  ( filterIfVisible()) );
+    setIconSize(QSize(32,32));
```
Comment (shundhammer) [inline src/YQPkgServiceList.cc]: Do we really have icons here?

File: src/YQPkgServiceList.cc [hunk 12]
```diff
@@ -0,0 +1,250 @@
...
+    QListIterator<QTreeWidgetItem *> it(items);
+
+    while ( it.hasNext() )
+    {
+      item = it.next();
+      YQPkgServiceListItem * serviceItem = dynamic_cast<YQPkgServiceListItem *> (item);
+
+        if ( serviceItem )
+        {
+            yuiMilestone() << "Selected service: " << serviceItem->zyppService() << std::endl;
```
Comment (shundhammer) [inline src/YQPkgServiceList.cc]: NP: You have `using std::endl` at the start of the file, so just `endl` is sufficient (and slightly better readable). Same below.

File: src/YQPkgServiceList.h [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+namespace zypp {
+    class RepoManager;
+}
+
+// just a service name
+typedef std::string	ZyppService;
+
+
+/**
+ * @short Display a list of libzypp services.
```
Comment (shundhammer) [inline src/YQPkgServiceList.h]: NP: Not quite correct. It should be "Widget to display a list of libzypp services". It's a class, not a method.
</wiseass_mode>

:smiling_imp:

File: src/YQPkgServiceList.cc (hunk 10 above)
Comment (lslezak) [reply src/YQPkgServiceList.cc]: I see your point, but I think it's not needed here. The `Name` column has always 100% width, see the screenshot at the top, the service names are usually long (I'd even say too long).

Moreover the same header is used in the "Repositories" view, we would need to adapt also the other places so it looks the same. And that might potentially break openQA so I'd keep it as it is.

File: src/YQPkgSecondaryFilterView.cc (hunk 7 above)
Comment (lslezak) [reply src/YQPkgSecondaryFilterView.cc]: Um, GitHub does not display indentation right when there is a mixture of tab and spaces, in my editor it looks good.

The only way would be converting tabs to spaces globally...

File: src/YQPkgSecondaryFilterView.cc (hunk 5 above)
Comment (lslezak) [reply src/YQPkgSecondaryFilterView.cc]: Added...

File: src/YQPkgSecondaryFilterView.cc (hunk 7 above)
Comment (shundhammer) [reply src/YQPkgSecondaryFilterView.cc]: OK then. I was just wondering.

File: src/YQPkgServiceList.cc [hunk 14]
```diff
@@ -194,16 +193,18 @@ YQPkgServiceListItem::YQPkgServiceListItem( YQPkgServiceList *	parentList,
...
     ZyppProduct product = singleProduct( _zyppService );
     if ( product )
     {
+        // TRANSLATORS: Tooltip item, followed by product name
         infoToolTip += ("<p><b>" + _("Product: ") + "</b>"
          + fromUTF8(product->summary()).toHtmlEscaped() +  "</p>");
     }
 
-    // list of the repositories included in the service
+    // TRANSLATORS: Tooltip item, followed by the list of repositories inluded in the libzypp sevice
```
Comment (shundhammer) [inline src/YQPkgServiceList.cc]: Typo: s/sevice/service

**Comment by shundhammer** [comment]
Just that one typo, otherwise LGTM.

**Pull request merged**